diff --git a/include/exec/fork_join.hpp b/include/exec/fork_join.hpp
index c9aa3421..436b70e4 100644
--- a/include/exec/fork_join.hpp
+++ b/include/exec/fork_join.hpp
@@ -293,6 +293,9 @@ namespace experimental::execution::__fork_join
 
     static constexpr auto __connect =
       []<class _Receiver, class _Sender>(_Sender&& __sndr, _Receiver&& __rcvr) noexcept
+      -> fork_join_impl_t::_opstate_t<STDEXEC::__child_of<_Sender>,
+                                      STDEXEC::__data_of<_Sender>,
+                                      _Receiver>
     {
       using _closures_t = STDEXEC::__data_of<_Sender>;
       using _sndr_t     = STDEXEC::__child_of<_Sender>;
